modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordConfigDropDown.kt

```diff
@@ -1,197 +1,138 @@
-import com.intellij.psi.search.GlobalSearchScope
+import com.intellij.openapi.components.service
+import com.intellij.openapi.fileEditor.FileEditorManager
+import com.intellij.openapi.project.ProjectLocator
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.openapi.vfs.VirtualFileManager
 import com.intellij.util.indexing.*
 import com.intellij.util.io.EnumeratorStringDescriptor
 import com.intellij.util.io.KeyDescriptor
-import java.nio.file.Path
-import java.util.*
+import java.util.Collections
 import javax.swing.JComponent
 
-// [`MirrordConfigIndex`] index is queried once in the update function to initialize the configFiles
-// Once initialized, we listen on file events to update the configFiles by querying the index again
-// If no config files are present, the chosenFile is set to null, the action is hidden,
-// and on pressing the gear icon, a file will be created
-// If one file is present, the chosenFile is set to that file, and the action is hidden, it is set as default
-// If two or more files are present, the chosenFile is set to the first file, and the action is shown
-// Startup -> MirrordConfigIndex -> (MirrordConfigWatcher -> Event -> Query Index -> Update Configs) -> MirrordConfigDropDown
-//                                    \<--------------------------------------------------------/
 class MirrordConfigDropDown : ComboBoxAction() {
+    private class EnableMirrordAction(val enabled: Boolean) : AnAction(if (enabled) "Enabled" else "Disabled") {
+        override fun actionPerformed(e: AnActionEvent) {
+            val service = e.project?.service<MirrordProjectService>() ?: return
+            service.enabled = !enabled
+        }
+    }
 
-    private lateinit var configFiles: HashSet<String>
+    private class ShowActiveConfigAction(val config: VirtualFile) : AnAction("Active Config: ${config.presentableUrl}") {
+        override fun actionPerformed(e: AnActionEvent) {
+            val service = e.project?.service<MirrordProjectService>() ?: return
+            FileEditorManager.getInstance(service.project).openFile(config, true)
+        }
+    }
 
-    @Volatile
-    private var blockQueries: Boolean = false
+    private class SelectActiveConfigAction : AnAction("Select Active Config") {
+        override fun actionPerformed(e: AnActionEvent) {
+            val service = e.project?.service<MirrordProjectService>() ?: return
 
-    override fun getActionUpdateThread() = ActionUpdateThread.BGT
+            val fileManager = VirtualFileManager.getInstance()
+            val projectLocator = ProjectLocator.getInstance()
+            val configs = FileBasedIndex
+                .getInstance()
+                .getAllKeys(MirrordConfigIndex.key, service.project)
+                .mapNotNull { fileManager.findFileByUrl(it) }
+                .filter { projectLocator.getProjectsForFile(it).contains(service.project) }
+                .associateBy { it.presentableUrl }
```

Comment:
i see you have not considered race conditions here/ states where we could have bad data.

querying the index in index mode is not good, we have an implementation in the main branch for this, so i suggest you change the querying to follow that approach:

```
    private fun blockAndQueryIndex(project: Project) {
        if (updateConfigs && !blockQueries) {
            blockQueries = true
            queryIndexInSmartMode(project, ::updateConfigFiles)
        }
    }
    ```

Reply:
Do you mean dumb mode? If so, this is already handled by the platform. The while popup is enabled only in the smart mode

Reply:
Can you please elaborate? How is handled already?

Reply:
I think for the action to be enabled in dumb mode you have to explicitly implement the `DumbAware` interface

Reply:
That’s just for that action, querying index is separate

Reply:
Yup

Reply:
After some reading - I think that's not possible, since updating indices is done in a `WriteAction` and this action has read access.

Reply:
Okay so as far as I can remember update is called every half second even when the action is unavailable. So I’d say check we don’t do anything related to keeping state of configs while indexing is on